web/html/src/components/table/AsyncDataHandler.js

```diff
@@ -0,0 +1,318 @@
+// @flow
+import * as React from 'react';
+
+import {Utils} from "utils/functions";
+import {ApiDataEndpoint, PageControl} from '../../utils/data-endpoint';
+import { StatePersistedContext } from '../utils/StatePersistedContext';
+
+import { ItemsPerPageSelector } from '../pagination';
+import { PaginationBlock } from '../pagination';
+import { SearchPanel } from './SearchPanel';
+import { SearchField } from './SearchField';
+import { Header } from './Header';
+
+type ChildrenArgsProps = {
+  currItems: Array<any>,
+  headers: React.Node,
+  handleSelect: Function,
+  selectable: boolean,
+  selectedItems: Array<any>,
+  criteria?: string,
+}
+
+type Props = {
+  columns: Array<React.ElementRef<any>>,
+  /** the async resource endpoint to retrieve the paged data from */
+  resource: string,
+  /** Function extracting the unique key of the row from the data object */
+  identifier: Function,
+  /** the column key name of the initial sorted column */
+  initialSortColumnKey?: string,
+  /** 1 for ascending, -1 for descending */
+  initialSortDirection?: number,
+  /** the React Object that contains the filter search field */
+  searchField?: React.Element<typeof SearchField>,
+  /** the initial number of how many row-per-page to show. If it's 0 table header and footer are hidden */
+  initialItemsPerPage?: number,
+  /** enables item selection */
+  selectable: boolean,
+  /** the handler to call when the table selection is updated. If not provided, the select boxes won't be rendered */
+  onSelect?: (items: Array<any>) => void,
+  /** the identifiers for selected items */
+  selectedItems?: Array<any>,
+  /** The message which is shown when there are no rows to display */
+  emptyText?: string,
+  /** The message which is shown when the data is loading */
+  loadingText?: string,
+  /** Children node in the table */
+  children: (args: ChildrenArgsProps) => React.Node,
+};
+
+type State = {
+  data: Array<any>,
+  currentPage: number,
+  itemsPerPage: number,
+  totalItems: number,
+  criteria?: string,
+  sortColumnKey: string | null,
+  sortDirection: number,
+  selectedItems: Array<any>,
+  selectable: boolean,
+  loading: boolean,
+  endpoint: ApiDataEndpoint
+};
```

Comment:
Most of these are common with `TableDataHandler`. I would rather see the `TableDataHandler` code refactored to accept something like a `dataProvider` property pointing to another component responsible for getting the data either asynchronously with your code or using preprovided data as we already have.